File: python3/Hash Maps and Sets/zero_striping_hash_sets.py
from typing import List


def zero_striping_hash_sets(matrix: List[List[int]]) -> None:     
    if not matrix or not matrix[0]:
        return
    m, n = len(matrix), len(matrix[0])
    zero_rows, zero_cols = set(), set()
    # Pass 1: Traverse through the matrix to identify the rows and
    # columns containing zeros and store their indexes in the 
    # appropriate hash sets.
    for r in range(m):
        for c in range(n):
            if matrix[r][c] == 0:
                zero_rows.add(r)
                zero_cols.add(c)
    # Pass 2: Set any cell in the matrix to zero if its row index is 
    # in 'zero_rows' or its column index is in 'zero_cols'.
    for r in range(m):
        for c in range(n):
            if r in zero_rows or c in zero_cols:
                matrix[r][c] = 0

File: go/Hash Maps and Sets/zero_striping_hash_sets.go
func zeroStripingHashSets(matrix [][]int) {
	if len(matrix) == 0 || len(matrix[0]) == 0 {
		return
	}
	m, n := len(matrix), len(matrix[0])
	zeroRows := make(map[int]struct{})
	zeroCols := make(map[int]struct{})
	// Pass 1: Traverse through the matrix to identify the rows and
	// columns containing zeros and store their indexes in the
	// appropriate hash sets.
	for r := 0; r < m; r++ {
		for c := 0; c < n; c++ {
			if matrix[r][c] == 0 {
				zeroRows[r] = struct{}{}
				zeroCols[c] = struct{}{}
			}
		}
	}
	// Pass 2: Set any cell in the matrix to zero if its row index is
	// in 'zero_rows' or its column index is in 'zero_cols'.
	for r := 0; r < m; r++ {
		for c := 0; c < n; c++ {
			if _, ok := zeroRows[r]; ok {
				matrix[r][c] = 0
			} else if _, ok := zeroCols[c]; ok {
				matrix[r][c] = 0
			}
		}
	}
}
